### PR TITLE
[Feature] add Tensor Parallelism to SD_3.5

### PR DESCRIPTION
**Pull request by GG-li**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Feature for https://github.com/vllm-project/vllm-omni/issues/1217.
Add Tensor Parallelism to Stable Diffusion-3.5.
## Test Plan
python text_to_image.py     --model stabilityai/stable-diffusion-3.5-medium     --prompt "a cup of coffee on the table"     --negative_prompt "ugly, unclear"     --cfg_scale 4.0     --num_inference_steps 50     --output "tp_enabled.png" --tensor_parallel_size 4
## Test Result
Image size: 1024x1024
| TP Size | Time | Generated Image |
| :--- | :--- | :--- |
| TP=1 | 5400.104ms | Result 1 |
| TP=2 |  2733.772ms | Result 2 |
| TP=4 |  2286.651ms | Result 4 |
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by GG-li** [comment]
@wtomin @Bounty-hunter Please take a look

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py [hunk 1]
```diff
@@ -18,6 +23,56 @@
...
+            raise ValueError(f"Unsupported activation function type: {activation_fn}")
+
+        self.net = nn.ModuleList([])
+        self.net.append(act_fn)
+        self.net.append(nn.Dropout(dropout))
+        self.net.append(RowParallelLinear(inner_dim, dim_out, bias=bias))
+        if final_dropout:
+            self.net.append(nn.Dropout(dropout))
+
+    def forward(self, hidden_states: torch.Tensor, *args, **kwargs) -> torch.Tensor:
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/sd3/sd3_transformer.py:68]: maybe to can rewrite as follow:
for layer in self.net:
    output = layer(hidden_states)
    hidden_states = output[0] if isinstance(output, tuple) else output
return hidden_states

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py [hunk 2]
```diff
@@ -18,6 +23,56 @@
...
+        dim_out = dim_out if dim_out is not None else dim
+
+        if activation_fn == "gelu-approximate":
+            act_fn = GELU(dim, inner_dim, approximate="tanh", bias=bias)
+        else:
+            raise ValueError(f"Unsupported activation function type: {activation_fn}")
+
+        self.net = nn.ModuleList([])
+        self.net.append(act_fn)
+        self.net.append(nn.Dropout(dropout))
```
Comment (wtomin) [inline vllm_omni/diffusion/models/sd3/sd3_transformer.py:63]: Can you explicilt set the `Dropout(p=0.0)` ? Just in case the training mode is on.

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py [hunk 3]
```diff
@@ -18,6 +23,56 @@
...
+        self.net.append(RowParallelLinear(inner_dim, dim_out, bias=bias))
+        if final_dropout:
+            self.net.append(nn.Dropout(dropout))
```
Comment (wtomin) [inline vllm_omni/diffusion/models/sd3/sd3_transformer.py:66]: Same `Dropout(p=0.0)` here.

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py [hunk 4]
```diff
@@ -18,6 +23,56 @@
...
+
+        if activation_fn == "gelu-approximate":
+            act_fn = GELU(dim, inner_dim, approximate="tanh", bias=bias)
+        else:
+            raise ValueError(f"Unsupported activation function type: {activation_fn}")
+
+        self.net = nn.ModuleList([])
+        self.net.append(act_fn)
+        self.net.append(nn.Dropout(dropout))
+        self.net.append(RowParallelLinear(inner_dim, dim_out, bias=bias))
```
Comment (wtomin) [inline vllm_omni/diffusion/models/sd3/sd3_transformer.py:64]: Is `input_is_parallel=True` required for `RowParallelLinear`?

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py [hunk 5]
```diff
@@ -122,22 +179,23 @@ def forward(
...
 
         # Apply QK normalization
         img_query = self.norm_q(img_query)
         img_key = self.norm_k(img_key)
 
         if encoder_hidden_states is not None:
             # Compute QKV for text stream (context projections)
-            qkv, _ = self.add_kv_proj(encoder_hidden_states)
-            txt_query, txt_key, txt_value = qkv.chunk(3, dim=-1)
+            qkv_add, _ = self.add_kv_proj(encoder_hidden_states)
```
Comment (wtomin) [inline vllm_omni/diffusion/models/sd3/sd3_transformer.py]: `self.add_kv_proj` may not be initialized if `added_kv_proj_dim is None`.

Needs to run check before calling `self.add_kv_proj`.

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py [hunk 6]
```diff
@@ -88,25 +144,26 @@ def __init__(
...
             self.to_add_out = None
 
         if not pre_only:
             self.to_out = nn.ModuleList([])
-            self.to_out.append(ReplicatedLinear(self.inner_dim, self.dim, bias=out_bias))
+            self.to_out.append(RowParallelLinear(self.inner_dim, self.dim, bias=out_bias, input_is_parallel=True))
         else:
             self.to_out = None
 
         self.norm_added_q = RMSNorm(head_dim, eps=eps) if qk_norm else nn.Identity()
```
Comment (wtomin) [inline vllm_omni/diffusion/models/sd3/sd3_transformer.py:159]: Can you check if you can replace diffusers' RMSNorm by vllm's:
```python
from vllm.model_executor.layers.layernorm import RMSNorm
```
I guess vllm's `RMSNorm` is better, and it is commonly used in vllm-omni

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py [hunk 7]
```diff
@@ -77,7 +132,8 @@ def __init__(
             hidden_size=dim,
             head_size=self.head_dim,
             total_num_heads=num_heads,
-            disable_tp=True,
+            bias=True,
+            return_bias=True,
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/sd3/sd3_transformer.py]: Please check for:
The returned bias was not used subsequently, which could easily lead to misunderstanding.
if skip_bias_add is false,  bias returns is None, we don't need to return.

**Comment by wtomin** [comment]
@GG-li Can you also help to support the online serving text-to-image script with tensor parallel? Sofar the online serving script does not have the argument corresponding to tensor parallel size.

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py (hunk 2 above)
Comment (GG-li) [reply vllm_omni/diffusion/models/sd3/sd3_transformer.py:63]: The dropout parameter is already explicitly set to 0.0 by default in the __init__ signature.
```
class FeedForward(nn.Module):
    def __init__(
        self,
        dim: int,
        dim_out: int | None = None,
        mult: int = 4,
        dropout: float = 0.0,
        activation_fn: str = "gelu-approximate",
        final_dropout: bool = False,
        inner_dim=None,
        bias: bool = True,
    ):
```

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py (hunk 3 above)
Comment (GG-li) [reply vllm_omni/diffusion/models/sd3/sd3_transformer.py:66]: Same as above.

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py (hunk 4 above)
Comment (GG-li) [reply vllm_omni/diffusion/models/sd3/sd3_transformer.py:64]: input_is_parallel is True by default. I've fixed the implementation.

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py (hunk 5 above)
Comment (GG-li) [reply vllm_omni/diffusion/models/sd3/sd3_transformer.py]: fixed

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py (hunk 6 above)
Comment (GG-li) [reply vllm_omni/diffusion/models/sd3/sd3_transformer.py:159]: fixed

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py (hunk 1 above)
Comment (GG-li) [reply vllm_omni/diffusion/models/sd3/sd3_transformer.py:68]: fixed

File: vllm_omni/diffusion/models/sd3/sd3_transformer.py (hunk 7 above)
Comment (GG-li) [reply vllm_omni/diffusion/models/sd3/sd3_transformer.py]: fixed

**Comment by wtomin** [comment]
@hsliuustc0106 @ZJY0516 @SamitHuang I think it is ready to merge.

**Review by ZJY0516 — APPROVED** [review]
LGTM
